---
title: "BismarkAnalysis"
output: html_document
date: "2024-05-23"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
    knitr::opts_knit$set(root.dir = "/beegfs/scratch/ric.cosr/moscato.giosue/MIRKO_WGBS2/")
getwd()
setwd("/beegfs/scratch/ric.cosr/moscato.giosue/MIRKO_WGBS2/")
```

```{r}
dir.create(file.path("DEA_out"),showWarnings = FALSE)
```

```{r}
library(DESeq2)
library(gprofiler2)
library(ggplot2)
library(ggrepel)
#library(clusterProfiler)
#library(enrichplot)
library(DOSE) # needed to convert to enrichResult object
library(EnhancedVolcano)
library(fgsea)
library('org.Hs.eg.db')
library('org.Mm.eg.db')
library(dplyr)
library(stringr)
library(gridExtra)
library(readr)
library(openxlsx)
library(clusterProfiler)
library(edgeR)


```


```{r}
files <- data.frame(path=list.files("/beegfs/scratch/ric.cosr/moscato.giosue/MIRKO_WGBS2/fastq/trimmed/trimmomatic/mapped/bismark/",pattern = ".cov.gz",full.names = T,include.dirs=T,recursive = T))
files$names <- str_split(str_split(files$path,"//",simplify = T)[,2],"_p",simplify=T)[,1]                    

yall <- readBismark2DGE(files$path,sample.names=files$names)
```
```{r}
yall
```

```{r}
dim(yall)

```

```{r}
 table(yall$genes$Chr)

```

```{r}
yall <- yall[yall$genes$Chr!="chrM", ]
```

```{r}
TSS <-  nearestTSS(yall$genes$Chr, yall$genes$Locus, species="Mm")
yall$genes$EntrezID <- TSS$gene_id
yall$genes$Symbol <- TSS$symbol
yall$genes$Strand <- TSS$strand
yall$genes$Distance <- TSS$distance
yall$genes$Width <- TSS$width
head(yall$genes)
```

#4.8.3 Filtering and normalization
We now turn to statistical analysis of differential methylation. Our first analysis will be for
individual CpG loci.
CpG loci that have low coverage are removed prior to downstream analysis as they provide
little information for assessing methylation levels. We sum up the counts of methylated and
unmethylated reads to get the total read coverage at each CpG site for each sample:
```{r}
Methylation <- gl(2,1,ncol(yall), labels=c("Me","Un"))
Me <- yall$counts[, Methylation=="Me"]
Un <- yall$counts[, Methylation=="Un"]
Coverage <- Me + Un
head(Coverage)

```
As a conservative rule of thumb, we require a CpG site to have a total count (both methylated
and unmethylated) of at least 8 in every sample before it is considered in the study.
```{r}
HasCoverage <- rowSums(Coverage >= 8) == 4

```

This filtering criterion could be relaxed somewhat in principle but the number of CpGs kept
in the analysis is large enough for our purposes.
We also filter out CpGs that are never methylated or always methylated as they provide no
information about differential methylation:
```{r}
 HasBoth <- rowSums(Me) > 0 & rowSums(Un) > 0
table(HasCoverage, HasBoth)
```
The DGEList object is subsetted to retain only the non-filtered loci:
```{r}
 y <- yall[HasCoverage & HasBoth,, keep.lib.sizes=FALSE]
```

A key difference between BS-seq and other sequencing data is that the pair of libraries
holding the methylated and unmethylated reads for a particular sample are treated as a unit.
To ensure that the methylated and unmethylated reads for the same sample are treated on
the same scale, we need to set the library sizes to be equal for each pair of libraries. We set
the library sizes for each sample to be the average of the total read counts for the methylated
and unmethylated libraries:
```{r}
TotalLibSize <- 0.5 * y$samples$lib.size[Methylation=="Me"] +
+ 0.5 * y$samples$lib.size[Methylation=="Un"]
y$samples$lib.size <- rep(TotalLibSize, each=2)
y$samples
```

#4.8.4 Data exploration
The data can be explored by generating multi-dimensional scaling (MDS) plots on the methylation level (M-value) of the CpG sites. The M-value is calcualted by the log of the ratio of methylated and unmethylated C’s, which is equivalent to the difference between methylated
and unmethylated C’s on the log-scale [13]. A prior count of 2 is added to avoid logarithms of zero.
```{r}
Me <- y$counts[, Methylation=="Me"]
Un <- y$counts[, Methylation=="Un"]
M <- log2(Me + 2) - log2(Un + 2)
colnames(M) <- files$names
```

Here M contains the empirical logit methylation level for each CpG site in each sample. We
have used a prior count of 2 to avoid logarithms of zero.
Now we can generate a multi-dimensional scaling (MDS) plot to explore the overall differences
between the methylation levels of the different samples.
```{r}
plotMDS(M, col=rep(1:3, each=2), main="M-values")
```
#4.8.5 Design matrix
One aim of this study is to identify differentially methylated (DM) loci between the different
cell populations. In edgeR, this can be done by fitting linear models under a specified design
matrix and testing for corresponding coefficients or contrasts. A basic sample-level design
matrix can be made as follows:
```{r}
files$Group <- c("NEU","NEU","NPC","NPC")
designSL <- model.matrix(~0+Group, data=files)
designSL
```
The we expand this to the full design matrix modeling the sample and methylation effects:
```{r}
design <- modelMatrixMeth(designSL)
design

```

#4.8.6 Differential methylation analysis at CpG loci
Then we proceed to testing for differentially methylated CpG sites between different groups.
We fit quasi NB GLM for all the CpG loci using the glmQLFit function.
```{r}
y1 <- y
y1 <- estimateDisp(y1,design)
fit <- glmQLFit(y1, design)
```

We identify differentially methylated CpG loci between the 40-45 and 60-65µm group using
the likelihood-ratio test. The contrast corresponding to this comparison is constructed using
the makeContrasts function.

```{r}
contr <- makeContrasts(GroupNEUvsNPC = GroupNEU  - GroupNPC, levels=design)
qlf <- glmQLFTest(fit, contrast=contr)
```

The top set of most significant DMRs can be examined with topTags. Here, positive log-fold
changes represent CpG sites that have higher methylation level in the NEu group compared to the NPC group. Multiplicity correction is performed by applying the BenjaminiHochberg method on the p-values, to control the false discovery rate (FDR).
```{r}
topTags(qlf)
```

The total number of DMRs in each direction at a FDR of 5% can be examined with decide
Tests.
```{r}
summary(decideTests(qlf))
```

The differential methylation results can be visualized using an MD plot. The difference of
the M-value for each CpG site is plotted against the average abundance of that CpG site.
Significantly DMRs at a FDR of 5% are highlighted.
```{r}
plotMD(qlf)
```

#4.8.7 Summarizing counts in promoter regions
It is usually of great biological interest to examine the methylation level within the gene
promoter regions. For simplicity, we define the promoter of a gene as the region from 2kb
upstream to 1kb downstream of the transcription start site of that gene. We then subset the
CpGs to those contained in a promoter region.
```{r}
InPromoter <- yall$genes$Distance >= -1000 & yall$genes$Distance <= 2000
yIP <- yall[InPromoter,,keep.lib.sizes=FALSE]
```

We compute the total counts for each gene promoter:
```{r}
> ypr <- rowsum(yIP, yIP$genes$EntrezID, reorder=FALSE)
> ypr$genes$EntrezID <- NULL
```

The integer matrix ypr$counts contains the total numbers of methylated and unmethylated
CpGs observed within the promoter of each gene.
Filtering is performed in the same way as before. We sum up the read counts of both
methylated and unmethylated Cs at each gene promoter within each sample
```{r}
> Mepr <- ypr$counts[,Methylation=="Me"]
> Unpr <- ypr$counts[,Methylation=="Un"]
> Coveragepr <- Mepr + Unpr
```

Since each row represents a 3,000-bps-wide promoter region that contains multiple CpG sites,
we would expect less filtering than before.
```{r}
> HasCoveragepr <- rowSums(Coveragepr >= 8) == 6
> HasBothpr <- rowSums(Mepr) > 0 & rowSums(Unpr) > 0
> table(HasCoveragepr, HasBothpr)
```

Same as before, we do not perform normalization but set the library sizes for each sample to
be the average of the total read counts for the methylated and unmethylated libraries.
```{r}
> TotalLibSizepr <- 0.5 * ypr$samples$lib.size[Methylation=="Me"] +
+ 0.5 * ypr$samples$lib.size[Methylation=="Un"]
> ypr$samples$lib.size <- rep(TotalLibSizepr, each=2)
> ypr$samples
```

#4.8.8 Differential methylation in gene promoters
We fit quasi NB GLMs for all the gene promoters using glmQLFit.
```{r}
> fitpr <- glmQLFit(ypr, design)

```

Then we can proceed to testing for differential methylation in gene promoter regions between
different populations. Suppose the comparison of interest is the same as before. The same
contrast can be used for the testing.
```{r}
> qlfpr <- glmQLFTest(fitpr, contrast=contr)

```


The top set of most differentially methylated gene promoters can be viewed with topTags:
```{r}
> topTags(qlfpr, n=20)

```


The total number of DM gene promoters identified at an FDR of 5% can be shown with
decideTests
```{r}
> summary(decideTests(qlfpr))

```

The differential methylation results can be visualized with an MD plot.
```{r}
> plotMD(qlfpr)

```





















